### PR TITLE
artifactory-client: add support for authorized range requests

### PR DESCRIPTION
**Pull request by origin-yaropolk**

### CODE REVIEW

File: src/utils/http.ts [hunk 1]
```diff
@@ -79,7 +84,7 @@ export async function request(url: string, method: string, requestData?: Request
 	return buffer ? buffer : Buffer.from('');
 }
 
-export function get(url: string, requestData?: RequestData): Promise<stream.Readable> {
+export function get(url: string, requestData?: RequestData): Promise<IncomingMessage> {
```
Comment (origin-yaropolk) [inline src/utils/http.ts:86]: There is no reason to reduce interface down to `Readable` here

File: src/artifactory-client.ts [hunk 2]
```diff
@@ -93,20 +93,30 @@ class Artifactory implements ArtifactoryClient {
...
+		}
+
+		return http.get(uri, reqData);
 	}
 
 	resolveUri(item: ArtifactoryItemMeta | string): string {
 		const baseUrl = `${this.config.protocol ?? 'https'}://${this.config.host}/artifactory/`;
 
 		if (typeof item === 'string') {
+			if (item.indexOf(baseUrl) === 0) {
```
Comment (origin-yaropolk) [inline src/artifactory-client.ts]: For case when _item_ has `baseUrl` already

File: src/artifactory-client.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
-import * as stream from 'stream';
 import * as http from './utils/http.js';
+import { IncomingMessage } from 'http';
```
Comment (Nipheris) [inline src/artifactory-client.ts]: `import type`?

File: src/artifactory-client.ts [hunk 4]
```diff
@@ -34,7 +34,7 @@ export interface AqlRequestResult<T> {
 
 export interface ArtifactoryClient {
 	query<T>(request: string): Promise<AqlRequestResult<T>>;
-	getContentStream(item: ArtifactoryItemMeta | string): Promise<stream.Readable>;
+	getContentStream(item: ArtifactoryItemMeta | string, rangeHeader?: string): Promise<IncomingMessage>;
```
Comment (Nipheris) [inline src/artifactory-client.ts]: I think we could pass `byteRange: ByteRange`, where `ByteRange` type is `{ start?: number; end?: number }` and build header in the client. Now it looks like a leaky abstraction: `rangeHeader` string should adhere to HTTP specification (i.e. it should be valid range string, like `bytes=10-100` or `bytes=-999`).

**Review by Nipheris — CHANGES_REQUESTED** [review]

File: src/artifactory-client.ts (hunk 4 above)
Comment (origin-yaropolk) [reply src/artifactory-client.ts]: Yep, it accepts raw header. Plan was that correct header would be formed in jewel-case side in `artifact-provider`.

File: src/artifactory-client.ts [hunk 5]
```diff
@@ -93,20 +93,30 @@ class Artifactory implements ArtifactoryClient {
...
+
+		if (rangeHeader && reqData.headers) {
+			reqData.headers.Range = rangeHeader;
+		}
+
+		return http.get(uri, reqData);
 	}
 
 	resolveUri(item: ArtifactoryItemMeta | string): string {
 		const baseUrl = `${this.config.protocol ?? 'https'}://${this.config.host}/artifactory/`;
```
Comment (HardParadox) [inline src/artifactory-client.ts]: why do we need to create `baseUrl` every time when we call this method? Can we move this logic to constructor?

Comment (HardParadox) [reply src/artifactory-client.ts]: and , I think, need to some refactor logic in this method

```suggestion
resolveUri(item: ArtifactoryItemMeta | string): string {
		if (typeof item !== 'string') {
				return `${this.baseUrl}${item.repo}/${item.path}/${item.name}`;
		}
		
		if (item.indexOf(baseUrl) === 0) {
				return item;
		}
		
		return `${this.baseUrl}${item}`;
```

File: src/artifactory-client.ts [hunk 6]
```diff
@@ -34,10 +34,15 @@ export interface AqlRequestResult<T> {
 
 export interface ArtifactoryClient {
 	query<T>(request: string): Promise<AqlRequestResult<T>>;
-	getContentStream(item: ArtifactoryItemMeta | string): Promise<stream.Readable>;
+	getContentStream(item: ArtifactoryItemMeta | string, byteRange?: ByteRange): Promise<IncomingMessage>;
 	resolveUri(item: ArtifactoryItemMeta | string): string;
 }
 
+export type ByteRange = {
```
Comment (HardParadox) [inline src/artifactory-client.ts]: this isn't very important, but usually  `interface` is used for such situations. And I think its place is above interface `ArtifactoryClient`

Comment (Nipheris) [reply src/artifactory-client.ts]: `ByteRange` is more like a tuple, not some interface to be implemented, so for me it is more natural to use `type`s in such cases)

**Review by HardParadox — APPROVED** [review]

**Review by Nipheris — APPROVED** [review]

**Pull request merged**